src/modules/loki/main.cpp

```diff
@@ -11,7 +11,7 @@
 int main(int argc, char *argv[]) {
   MainProg Main("loki", argc, argv);
 
-  auto Detector = new Caen::CaenBase(Main.DetectorSettings);
-
+  auto Detector = new Caen::CaenBase(Main.DetectorSettings, ESSReadout::Parser::LOKI);
```

Comment:
This is a good idea, but I kind of get the feeling that passing through both detectorsettings and data type is semantically redundant. When the detector name is already given in the MainProg constructor maybe we can just use that information? or could there be a reason to want to be able to decouple these?

Then in config we can use a factory pattern to both update settings and set the detector type. Then at least this is only done in one place? So something like this:

In stead of:
if (settings.DetectorName == "loki") {
    Geom = new LokiGeometry(CaenConfiguration);
  } else if (settings.DetectorName == "bifrost") {
    Geom = new BifrostGeometry(CaenConfiguration);
  } else if (settings.DetectorName == "miracles") {
    Geom = new MiraclesGeometry(CaenConfiguration);
  } else {
    XTRACE(INIT, ERR, "Invalid Detector Name %s",
           settings.DetectorName.c_str());
    throw std::runtime_error(
        fmt::format("Invalid Detector Name {}", settings.DetectorName));
  }
  
  we could do something like:
  
  Geom = InstrumentFactory.getGeometry(CaenConfiguration)
  Type = InstrumentFactory.getDataType(CaenConfiguration)

Reply:
I like this idea, I'll merge this in so that Greg can use master again without any issues, and will make a jira ticket to implement this